pennylane/devices/device_interface.py

```diff
@@ -0,0 +1,399 @@
+        >>> python_device.supports_gradient_with_configuration(config)
+        True
+        >>> cpp_device.supports_gradient_with_configuration(config)
+        False
+
+        """
+        if getattr(execution_config, "diff_method", None) == "backprop":
+            return False
+        order = getattr(execution_config, "order", 1)
+        return cls.gradient != AbstractDevice.gradient if order == 1 else False
+
+    def gradient(
+        self,
+        qscript: QScript_or_QScriptBatch,
+        execution_config=None,
+    ):
+        """Calculate the gradient of either a single or a batch of Quantum Scripts.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            Tuple: The gradient for each trainable parameter
+
+        See also :meth:`~.supports_gradient_with_configuration` and :meth:`~.execute_and_gradients`.
+
+        **Execution Config:**
+
+        The execution config has an ``order`` property that describes the order of gradient requested. If the requested order
+        of gradient is not provided, the device should raise a ``NotImplementedError``. The :meth:`~.supports_gradient_with_configuration`
+        method can pre-validate supported orders.
+
+        **Return Shape:**
+
+        If a batch of quantum scripts is provided, this method should return a tuple with each entry being the gradient of
+        each individual quantum script. If the batch is of length 1, then the return tuple should still be of length 1, not squeezed.
+
+        """
+        raise NotImplementedError
+
+    def execute_and_gradients(
+        self,
+        qscripts: QScript_or_QScriptBatch,
+        execution_config=None,
+    ):
+        """Compute the results and gradients of QuantumScripts at the same time.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript or batch to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            tensor-like, tensor-like: A numeric result of the computation and the gradient.
+
+        See :meth:`~.execute` and :meth:`~.gradient` for more information about return shapes and behaviour. If :meth:`~.gradient`
+        is defined, this method should be as well.
+
+        This method can be used when the result and execution need to be computed at the same time, such as
+        during a forward mode calculation of gradients. For certain gradient methods, such as adjoint
+        diff gradients, calculating the result and gradient at the same can save computational work.
+
+        """
+        return self.execute(qscripts, execution_config), self.gradient(qscripts, execution_config)
```

Comment:
Just to understand, this is just the default, and a device that has a more efficient way of computing both would just overwrite? (I assume almost any device should have a more efficient way :) )

Reply:
Exactly. For example, our python device could have `execute` return auxiliary information containing the "prerotated" state that would then be used for adjoint differentiation.